lib/sycamore/sycamore/writers/base.py

```diff
@@ -0,0 +1,160 @@
+        _check_serializable(datasink)
+        record_dataset.write_datasink(datasink, ray_remote_args=self.resource_args)
+        return input_dataset
+
+    def _ray_map_docs_to_records(self, dataset: Dataset) -> Dataset:
+
+        def build_ray_callable():
+            cls = self.__class__
+
+            def ray_callable(ray_input: dict[str, np.ndarray]) -> dict[str, list]:
+                all_docs = [Document.deserialize(s) for s in ray_input.get("doc", [])]
+                docs = [d for d in all_docs if not isinstance(d, MetadataDocument)]
+                records = [cls.doc_to_record(d) for d in docs]
+                return {"record": [r.serialize() for r in records]}
+
+            return ray_callable
+
+        ray_callable = build_ray_callable()
+        _check_serializable(ray_callable)
+
+        return dataset.map_batches(ray_callable, **self.resource_args)
+
+    class InnerDatasink(Datasink):
+
+        def __init__(
+            self,
+            client_params: "BaseDBWriter.client_params_t",
+            index_params: Optional["BaseDBWriter.index_params_t"],
+            owner_cls: type["BaseDBWriter"],
+        ):
+            _check_serializable(client_params, index_params, owner_cls)
+
+            self._client_params = client_params
+            self._index_params = index_params
+            self._owner = owner_cls
+
+        def on_write_start(self) -> None:
+            if self._index_params:
+                client = self._owner.client_t.from_client_params(self._client_params)
+                client.create_index_if_missing(self._index_params)
+
+        def write(self, blocks: Iterable[Block], ctx: TaskContext) -> Any:
+            builder = DelegatingBlockBuilder()
+            for block in blocks:
+                builder.add_block(block)
+            master_block = builder.build()
+            rows = BlockAccessor.for_block(master_block).to_arrow().to_pylist()
+            records = [self._owner.record_t.deserialize(row["record"]) for row in rows]
+            client = self._owner.client_t.from_client_params(self._client_params)
+            client.write_many_records(records)
+
+
+class BaseMetadataDBWriter(BaseDBWriter):
+
+    def _ray_map_docs_to_records(self, dataset: Dataset) -> Dataset:
+
+        def build_ray_callable():
+            cls = self.__class__
+
+            def ray_callable(ray_input: dict[str, np.ndarray]) -> dict[str, list]:
+                all_docs = [Document.deserialize(s) for s in ray_input.get("doc", [])]
+                meta_docs = [d for d in all_docs if isinstance(d, MetadataDocument)]
+                records = [cls.doc_to_record(d) for d in meta_docs]
+                return {"record": [r.serialize() for r in records]}
+
+            return ray_callable
+
+        ray_callable = build_ray_callable()
+        _check_serializable(ray_callable)
+
+        return dataset.map_batches(ray_callable, **self.resource_args)
+
+
+def _check_serializable(*objects):
```

Comment:
can we move this into some shared class and use it in basemap also (maybe the map rewrite will mean you don't need this anymore, but I can't tell for sure)